references/debugging_guide.md

```diff
@@ -0,0 +1,95 @@
+
+1. Track the failure.
+   1. Save one's work and the current state of the system. Check in the saved state via git with a descriptive message using the following format "[Bug] Insert 73-character-or-less bug description here."
+   2. Create a github issue that documents the bug's existence. This will be a bug report containing:
+      - A one line description of the problem. Should basically be the same as the message from the previous step.
+      - A written description of the failure. What happened, in words?
+      - A written description of the expected behavior. What was supposed to happen, in words?
+      - Severity categorization: {Blocker; Major - aka large problem; Minor - aka an easy to fix problem; Enhancement - aka not a requirement but nice to have.}
+      - A minimal working example or set of steps that reproduces the failure.
+      - If possible (but not required now), a simplified test example that produces the same qualitative failure.
+      - Any diagnostic information (errors, execution logs, etc.) reported when the failure was observed.
+      - Environment information when the failure was observed (installed packages, operating system, memory capacity, etc.)
+   3. Create a new branch off the feature branch to track work on fixing the bug.
+   4. Open a pull request for the new branch to the original feature branch. Register that the issue is being worked on in this pull request by referencing the issue for the bug in the text body of the pull request on github.
+   5. Add a test to the suite of regression tests to reproduce this specific software failure instance. Check the test in with git.
+      - See chapter 4, "Reproducing Problems" of "Why Programs Fail: A Guide to Systematic Debugging" by Morgan Kaufmann.
+   6. Create an automated and simplified test case for this specific failure instance.
+      - See chapter 5, "Simplifying Problems" of "Why Programs Fail: A Guide to Systematic Debugging" by Morgan Kaufmann.
+
+2. Understand how the system that is producing the failure is supposed to work.
+   1. Identify and read the basic documentation related to the (relevant?) external software tools being used at or around the point of failure.
+   2. Take basic trainings (e.g. online tutorials, online courses, etc.) in the external tools being used in your project.
+   3. Identify, store, and read the details about major / complex algorithms that one's software is implementing.
+
+3. Understand how the system that is producing the failure is working.
+   1. What is being produced during failure?
+
+      Ensure that one has a debugging system in-place to identify and record the intermediate values being produced by one's software as well as which of those values are incorrect in the simplest created test case.
+
+   2. When does failure occur?
+
+      If failure is intermittent or conditional, document what inputs, if any, cause the failure to occur rather than not.
```

Comment:
I think this step could be merged with step 3.1.4. as follows: 

> Be sure to include logically likely contributors to the failure's existence (e.g. other known bugs, causes in program state / code / input, anomolies in the output or intermediate variables, poorly written code, etc.). If failure is intermittent or conditional, document what inputs, if any, cause the failure to occur rather than not.

Change suggestion for this change is below.

Reply:
Edited the doc to note the difference between recording the inputs that cause a failure and hypothesizing about why those inputs cause the failure or hypothesizing about where in the code the defect is.